### ISSUE TITLE
Expose StringReader.Position, ideally with an interface

### ISSUE THREAD

**Issue by loop-evgeny** [dotnet/runtime#53218]
## Background and Motivation

When reading from a TextReader it's sometimes useful to get its current position, e.g. to inform the user of progress when processing a long string. TextReader does not have a Position property, but there are only 2 BCL implementation: StreamReader and StringReader. With StreamReader the user can check `.BaseStream.Position`. StringReader stores its position in a private field (`_pos`).

Ideally there would be a `TextReader.Position` property, but that would, of course, be a huge breaking change. Just adding `StringReader.Position` should be a simple and non-breaking, though.

A nice "next step" would be to add an interface for it (`IPositionInfo`), which could be implemented by both `StringReader` and `StreamReader` (delegating to `BaseStream.Position`). Then user could could check for this interface instead of specific TextReader implementations and user-defined TextReader could also implement this.

## Proposed API

1)

```
namespace System.IO
{
    public class StringReader : TextReader
    {
        public long Position => _pos;
    }
}
```

2)

```
namespace System.IO
{
    public interface IPositionInfo
    {
        public long Position { get; }
    }

    public class StringReader : TextReader, IPositionInfo

    public class StreamReader : TextReader, IPositionInfo
}
```

## Usage Examples

``` C#
            StringReader stringReader = new StringReader(aVeryLongInputString);
            JsonReader jsonReader = new JsonTextReader(stringReader);

            while (jsonReader.Read())
            {
                // Do something with the JSON
                Console.WriteLine($"{stringReader.Position * 100.0 / aVeryLongInputString.Length}% read");
            }
```

**Comment by FilipToth** [comment]
I don't think an interface is necessary in this case. Can you maybe tell us some advantages of using an interface in this particular case, probably I just don’t see it?

**Comment by filipnavara** [comment]
> With `StreamReader` the user can check `.BaseStream.Position`

That's actually inaccurate. `StreamReader` does buffering so the position in the underlying stream is usually way past of the number of data you have read. Not to mention that you can also read the stream from a position that doesn't start at zero.

**Comment by loop-evgeny** [comment]
The use case for an interface would be a method taking a TextReader, rather than a StringReader:

```
    void ProcessString() => ProcessText(StringReader(aVeryLongInputString));

    void ProcessText(TextReader textReader)
    {
            JsonReader jsonReader = new JsonTextReader(textReader);

            while (jsonReader.Read())
            {
                // Do something with the JSON
                if (textReader is IPositionInfo posInfo)
                {
                    Console.WriteLine($"{posInfo.Position * 100.0 / aVeryLongInputString.Length}% read");
                }
            }
    }
```

>> With StreamReader the user can check .BaseStream.Position
> That's actually inaccurate. StreamReader does buffering so the position in the underlying stream is usually way past the number of data you have read. Not to mention that you can also read the stream from a position that doesn't start at zero.

Sure, it's not perfect, but for my use case where you're processing a large input in its entirety (which I think would a common use case) it's good enough to be useful. Of course, it would be better if StreamReader provided an accurate Position property, too - but perhaps that should be a separate issue.

**Comment by filipnavara** [comment]
> it would be better if StreamReader provided an accurate Position property

...but position in what units? When reading stream it could be characters or bytes.

For characters it's something you can calculate yourself while reading it, just as you can do with `StringReader`. You can even make your own wrapper around `TextReader` that delegates to inner `TextReader` and tracks the position.

If the answer is bytes then I agree that's something that would actually be useful but it's something totally specific to `StreamReader` that would not apply to other `TextReader` implementations.

**Comment by loop-evgeny** [comment]
> For characters it's something you can calculate yourself while reading it, just as you can do with StringReader

How would I do that with StringReader? It doesn't expose anything, as far as I can see.

A TextReader wrapper is a possibility, yes, but it's quite a lot of code (and with that, the possibility of bugs).

**Comment by FilipToth** [comment]
> For characters it's something you can calculate yourself while reading it, just as you can do with StringReader (short of mixed line endings). You can even make your own wrapper around TextReader that delegates to inner TextReader and tracks the position.

Yes, you could also make your own `TextReader` class. But we're not doing that either. The question should be if we take the number of char/bytes read from the start of the document or just the line?

**Comment by filipnavara** [comment]
> Well, we can technically convert the byte length to char length so, I suppose bytes would be more useful.

You cannot convert it accurately. There's no guarantee that `Encoding` is symmetrical. `StringReader` never deals with bytes so it would be quite an overhead and unnatural extension on the class.

**Comment by loop-evgeny** [comment]
Characters would make sense to me as a unit - this is what TextReader deals with, e.g. `int Read(char[] buffer, int index, int count)` takes a count in characters.

So, to be clear, the primary request here is simply this, on StringReader: `public long Position => _pos;`

If StreamReader can also return its position in characters (which sounds likely, because it has `private int _charPos;`) then that would be a nice extra. Then it would make sense, IMO, to have a common interface through which code with a TextReader can read these without knowing what derived class it is. (A little bit like `IXmlLineInfo`, but for characters.)

**Comment by trivalik** [comment]
Right now I end up in requiring a Position. I am sad that `StreamReader` still has no way to read current position.

My workaround so far:

```
        public class ReaderWithPosition : StreamReader
        {
            private static readonly FieldInfo CurrentPositionField;

            static ReaderWithPosition()
            {
                CurrentPositionField = typeof(StreamReader).GetField("charPos", BindingFlags.NonPublic | BindingFlags.Instance) ?? typeof(StreamReader).GetField("_charPos", BindingFlags.NonPublic | BindingFlags.Instance); // make sure work for .NET Framework and .NET Core
            }
            
            // here add all necessary constructor

            public int CurrentPosition { get { return (int)CurrentPositionField.GetValue(this); } }
        }
```
